src/axom/primal/geometry/NURBSPatch.hpp

```diff
@@ -0,0 +1,2745 @@
+   *   uniquely determine the degree
+   * 
+   * \pre Requires a valid knot vector and npts_d > deg_d
+   */
+  NURBSPatch(const CoordsMat& pts,
+             const WeightsMat& weights,
+             const axom::Array<T>& knots_u,
+             const axom::Array<T>& knots_v)
+    : m_controlPoints(pts)
+    , m_weights(weights)
+  {
+    auto pts_shape = pts.shape();
+
+    SLIC_ASSERT(pts_shape[0] >= 0 && pts_shape[1] >= 0);
+
+    m_knotvec_u = KnotVectorType(knots_u, knots_u.size() - pts_shape[0] - 1);
+    m_knotvec_v = KnotVectorType(knots_v, knots_v.size() - pts_shape[1] - 1);
+
+    SLIC_ASSERT(isValidNURBS());
+  }
+
+  /*!
+   * \brief Constructor for a NURBS Patch from 1D axom::Array array of coordinates and KnotVector objects
+   *
+   * \param [in] pts A 2D axom::Array of (ord_u+1, ord_v+1) control points
+   * \param [in] knotvec_u A KnotVector object for the first axis
+   * \param [in] knotvec_v A KnotVector object for the second axis
+   * 
+   * For clamped and continuous curves, npts and the knot vector 
+   *   uniquely determine the degree
+   * 
+   * \pre Requires a valid knot vector and npts_d > deg_d
+   */
+  NURBSPatch(const CoordsMat& pts,
+             const KnotVectorType& knotvec_u,
+             const KnotVectorType& knotvec_v)
+    : m_controlPoints(pts)
+    , m_knotvec_u(knotvec_u)
+    , m_knotvec_v(knotvec_v)
+  {
+    makeNonrational();
+    SLIC_ASSERT(isValidNURBS());
+  }
+
+  /*!
+   * \brief Constructor for a NURBS Patch from 2D axom::Array array of coordinates, weights, and KnotVector objects
+   *
+   * \param [in] pts A 2D axom::Array of (ord_u+1, ord_v+1) control points
+   * \param [in] weights A 2D axom::Array of (ord_u+1, ord_v+1) positive weights
+   * \param [in] knotvec_u A KnotVector object for the first axis
+   * \param [in] knotvec_v A KnotVector object for the second axis
+   * 
+   * For clamped and continuous curves, npts and the knot vector 
+   *   uniquely determine the degree
+   * 
+   * \pre Requires a valid knot vector and npts_d > deg_d
+   */
+  NURBSPatch(const CoordsMat& pts,
+             const WeightsMat& weights,
+             const KnotVectorType& knotvec_u,
+             const KnotVectorType& knotvec_v)
+    : m_controlPoints(pts)
+    , m_weights(weights)
+    , m_knotvec_u(knotvec_u)
+    , m_knotvec_v(knotvec_v)
+  {
+    SLIC_ASSERT(isValidNURBS());
+  }
+
+  /*!
+   * \brief Evaluate a NURBS surface at a particular parameter value \a t
+   *
+   * \param [in] u The parameter value on the first axis
+   * \param [in] v The parameter value on the second axis
+   * 
+   * Adapted from Algorithm A3.5 on page 103 of "The NURBS Book"
+   * 
+   * \pre Requires \a u, v in the span of each knot vector
+   */
+  PointType evaluate(T u, T v) const
+  {
+    SLIC_ASSERT(u >= m_knotvec_u[0] &&
+                u <= m_knotvec_u[m_knotvec_u.getNumKnots() - 1]);
+    SLIC_ASSERT(v >= m_knotvec_v[0] &&
+                v <= m_knotvec_v[m_knotvec_v.getNumKnots() - 1]);
```

Comment:
This logic triggered an assert for me (in the NURBS curve class) when using a for loop with doubles, e.g.
```cpp
const double dt = 0.1
for(double u=0; u<=1.; u+=dt)
{
   ...
}
```

Would it make sense to check for fuzzy equals and clamp to the endpoints?